### PR TITLE
Add Equivalence Relation Definitions to Dafny Subfolder

### PR DESCRIPTION
**Pull request by stefanzetzsche**
Follow up to https://github.com/dafny-lang/libraries/pull/105

By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

### CODE REVIEW

**Review by keyboardDrummer — APPROVED** [review]

File: src/dafny/Relations.dfy [hunk 1]
```diff
@@ -26,6 +26,10 @@ module {:options "-functionSyntax:4"} Dafny.Relations {
     forall x, y :: R(x, y) && R(y, x) ==> x == y
   }
 
+  ghost predicate Symmetric<T(!new)>(R: (T, T) -> bool) {
+    forall x, y :: R(x, y) <==> R(y, x)
```
Comment (keyboardDrummer) [inline src/dafny/Relations.dfy:30]: Just a question for my own education, what's the difference between using `<==>` and `==` here?

File: src/dafny/Relations.dfy [hunk 2]
```diff
@@ -26,6 +26,10 @@ module {:options "-functionSyntax:4"} Dafny.Relations {
     forall x, y :: R(x, y) && R(y, x) ==> x == y
   }
 
+  ghost predicate Symmetric<T(!new)>(R: (T, T) -> bool) {
```
Comment (keyboardDrummer) [inline src/dafny/Relations.dfy:29]: What's the reason a predicate like this shouldn't be used for heap based types? (`!new`)

File: src/dafny/Relations.dfy (hunk 1 above)
Comment (stefan-aws) [reply src/dafny/Relations.dfy:30]: According to Section 5.2.1.1. in https://dafny.org/dafny/DafnyRef/DafnyRef, _the expressions `A <==> B` and `A == B` give the same value, but note that `<==>` is associative whereas `==` is chaining and they have different precedence_.

File: src/dafny/Relations.dfy (hunk 2 above)
Comment (stefan-aws) [reply src/dafny/Relations.dfy:29]: The reason is quantification: removing `!new` leads to the error: _a forall expression involved in a predicate definition is not allowed to depend on the set of allocated references, but values of 'x' (of type 'T') may contain references (perhaps declare its type as 'T(!new)') (see documentation for 'older' parameters)_.

**Pull request merged**